e2e/test/scenarios/models/reproductions.cy.spec.ts

```diff
@@ -444,3 +447,107 @@ describe("issue 41785", () => {
       .should("not.exist");
   });
 });
+
+describe.skip("issue 40635", () => {
+  beforeEach(() => {
+    restore();
+    cy.signInAsNormalUser();
+    cy.intercept("POST", "/api/dataset").as("dataset");
+  });
+
+  it("correctly displays question's and nested model's column names (metabase#40635)", () => {
+    startNewQuestion();
+    entityPickerModal().within(() => {
+      entityPickerModalTab("Tables").click();
+      cy.findByText("Orders").click();
+    });
+
+    getNotebookStep("data").button("Pick columns").click();
+    popover().findByText("Select none").click();
+
+    join();
+
+    entityPickerModal().within(() => {
+      entityPickerModalTab("Tables").click();
+      cy.findByText("Products").click();
+    });
+
+    getNotebookStep("join", { stage: 0, index: 0 })
+      .button("Pick columns")
+      .click();
+    popover().within(() => {
+      cy.findByText("Select none").click();
+      cy.findByText("ID").click();
+    });
+
+    join();
+
+    entityPickerModal().within(() => {
+      entityPickerModalTab("Tables").click();
+      cy.findByText("Products").click();
+    });
+
+    getNotebookStep("join", { stage: 0, index: 1 })
+      .button("Pick columns")
+      .click();
+    popover().within(() => {
+      cy.findByText("Select none").click();
+      cy.findByText("ID").click();
+    });
+
+    getNotebookStep("join", { stage: 0, index: 1 })
+      .findByText("Product ID")
+      .click();
+    popover().findByText("User ID").click();
+
+    visualize();
+    assertSettingsSidebar();
+    assertVisualizationColumns();
+
+    cy.button("Save").click();
+    modal().button("Save").click();
+    modal().findByText("Not now").click();
+
+    assertSettingsSidebar();
+    assertVisualizationColumns();
+
+    openQuestionActions();
+    popover().findByTextEnsureVisible("Turn into a model").click();
+    modal().button("Turn this into a model").click();
+    undoToast().should("contain", "This is a model now").icon("close").click();
+
+    assertSettingsSidebar();
+    assertVisualizationColumns();
+
+    openNotebook();
+    getNotebookStep("data").button("Pick columns").click();
+    popover().within(() => {
+      cy.findAllByText("ID").should("have.length", 1);
+      cy.findAllByText("Products → ID").should("have.length", 1);
+      cy.findAllByText("Products_2 → ID").should("have.length", 1);
```

Comment:
These assertions will fail. `Products → ID` won't be present, instead `Products_2 → ID` will be there twice.

Reply:
Ran the test locally and seeing exactly what you described.
![image](https://github.com/metabase/metabase/assets/31325167/1d43d5fa-7e5a-4296-b0b6-af246f528e44)

Reply:
Will do next time, thanks!